ocaml/xapi/message_forwarding.ml

```diff
@@ -3899,8 +3952,13 @@ module Forward = functor(Local: Custom_actions.CUSTOM_ACTIONS) -> struct
     let destroy ~__context ~self =
       info "PBD.destroy: PBD '%s'" (pbd_uuid ~__context self);
       let sr = Db.PBD.get_SR ~__context ~self in
+      let host = Db.PBD.get_host ~__context ~self in
       SR.with_sr_marked ~__context ~sr ~doc:"PBD.destroy" ~op:`pbd_destroy
-        (fun () -> Local.PBD.destroy ~__context ~self)
+        (fun () -> (* Forget metrics on the host which owns that SR. This is safe
+                      because on a host there is at most one pbd per sr *)
+                   do_op_on ~local_fn:(Local.SR.forget ~sr) ~__context ~host
+                     (fun session_id rpc -> Client.SR.forget rpc session_id sr);
+                   Local.PBD.destroy ~__context ~self)
```

Comment:
as far as i'm concerned this PR is working as expected and is good to go

however, i would like to get your opinion on changing the `PBD.destroy` api @robhoes ?

Reply:
What if this is a shared SR that I have detached from one host and did a PBD destroy on? Wouldn't forgetting the SR be problematic for the other hosts that are still using it?

Reply:
yes - i will test something like the following then

```
fun () -> if (shared_sr && i_am_last_host connected_to_sr) || not shared_sr then
                forget_sr ()
              pbd_destroy ()

```

Reply:
i have realised that `Client.SR.forget` and `Local.SR.forget` are actually a bit different, since `Client.SR.forget` performs in addition a `Xapi_sr.really_forget`, so I need to think quite carefully about what is going on

Reply:
i think the solution is to allow `SR.forget` when an SR has no PBDs, and for `PBD.destroy` to do what it did before

Reply:
@lippirk That sounds right. I recall talking about this a couple of weeks ago.

By the way, when you say `Client.SR.forget` and `Local.SR.forget` are different, do you mean that the former has things going on in `Message_forwarding` while the latter doesn't? That is not exactly true. In general, the `Client` module is for sending an call through the API layer to a xapi daemon on the same or another host. In `Message_forwarding`, this is normally used to forward a call from the master to a slave host. When a slave receives a forwarded API call, it does not go through message forwarding again, but the local function is called directly.

Reply:
yes that's what i meant

ah i see, i thought that we go through message forwarding again. so when an rpc is sent from one xapi to another it contains information about whether it has already been forwarded and the receiving xapi unpacks this accordingly into a `Local.X.Y` call?

Reply:
See https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/server_helpers.ml#L76.